r/aws_dx_gateway_association: Update for transit gateway support plus the second half (cross-account proposal acceptance) of support for Gateway Association Proposals

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

### Transit gateway support

Part of the work for https://github.com/terraform-providers/terraform-provider-aws/issues/8490.

### Cross-account proposal acceptance

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/8100.
Follows on from:
* https://github.com/terraform-providers/terraform-provider-aws/pull/8320
* https://github.com/terraform-providers/terraform-provider-aws/issues/8199

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
NOTES:

* resource/aws_dx_gateway_association: The `vpn_gateway_id` attribute is being deprecated in favor of the new `associated_gateway_id` attribute to support transit gateway associations (https://github.com/terraform-providers/terraform-provider-aws/issues/8490)
* resource/aws_dx_gateway_association_proposal: The `vpn_gateway_id` attribute is being deprecated in favor of the new `associated_gateway_id` attribute to support transit gateway associations (https://github.com/terraform-providers/terraform-provider-aws/issues/8490)

ENHANCEMENTS:

* resource/aws_dx_gateway_association: New attributes `associated_gateway_owner_account_id` and `proposal_id` added to support the acceptance of a Direct Connect gateway association proposal and create a cross-account Direct Connect gateway association (https://github.com/terraform-providers/terraform-provider-aws/issues/8100).
New attribute `associated_gateway_id` replaces deprecated `vpn_gateway_id` attribute to support transit gateway associations (https://github.com/terraform-providers/terraform-provider-aws/issues/8490).
New exported attributes `associated_gateway_type` and `dx_gateway_owner_account_id`
* resource/aws_dx_gateway_association_proposal: New attribute `associated_gateway_id` replaces deprecated `vpn_gateway_id` attribute to support transit gateway associations (https://github.com/terraform-providers/terraform-provider-aws/issues/8490).
New exported attributes `associated_gateway_owner_account_id` and `associated_gateway_type`
```

ewbankkit commented:
Acceptance/regression tests for tidy up of VGW attachments:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpnGatewayAttachment_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSVpnGatewayAttachment_ -timeout 120m
=== RUN   TestAccAWSVpnGatewayAttachment_basic
=== PAUSE TestAccAWSVpnGatewayAttachment_basic
=== RUN   TestAccAWSVpnGatewayAttachment_deleted
=== PAUSE TestAccAWSVpnGatewayAttachment_deleted
=== CONT  TestAccAWSVpnGatewayAttachment_basic
=== CONT  TestAccAWSVpnGatewayAttachment_deleted
--- PASS: TestAccAWSVpnGatewayAttachment_deleted (194.28s)
--- PASS: TestAccAWSVpnGatewayAttachment_basic (199.47s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	199.529s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpnGateway_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 2 -run=TestAccAWSVpnGateway_ -timeout 120m
=== RUN   TestAccAWSVpnGateway_importBasic
=== PAUSE TestAccAWSVpnGateway_importBasic
=== RUN   TestAccAWSVpnGateway_basic
=== PAUSE TestAccAWSVpnGateway_basic
=== RUN   TestAccAWSVpnGateway_withAvailabilityZoneSetToState
=== PAUSE TestAccAWSVpnGateway_withAvailabilityZoneSetToState
=== RUN   TestAccAWSVpnGateway_withAmazonSideAsnSetToState
=== PAUSE TestAccAWSVpnGateway_withAmazonSideAsnSetToState
=== RUN   TestAccAWSVpnGateway_disappears
=== PAUSE TestAccAWSVpnGateway_disappears
=== RUN   TestAccAWSVpnGateway_reattach
=== PAUSE TestAccAWSVpnGateway_reattach
=== RUN   TestAccAWSVpnGateway_delete
=== PAUSE TestAccAWSVpnGateway_delete
=== RUN   TestAccAWSVpnGateway_tags
=== PAUSE TestAccAWSVpnGateway_tags
=== CONT  TestAccAWSVpnGateway_importBasic
=== CONT  TestAccAWSVpnGateway_disappears
--- PASS: TestAccAWSVpnGateway_disappears (147.95s)
=== CONT  TestAccAWSVpnGateway_tags
--- PASS: TestAccAWSVpnGateway_importBasic (171.30s)
=== CONT  TestAccAWSVpnGateway_delete
--- PASS: TestAccAWSVpnGateway_tags (178.94s)
=== CONT  TestAccAWSVpnGateway_reattach
--- PASS: TestAccAWSVpnGateway_delete (162.02s)
=== CONT  TestAccAWSVpnGateway_withAvailabilityZoneSetToState
--- PASS: TestAccAWSVpnGateway_withAvailabilityZoneSetToState (159.52s)
=== CONT  TestAccAWSVpnGateway_withAmazonSideAsnSetToState
--- PASS: TestAccAWSVpnGateway_withAmazonSideAsnSetToState (136.88s)
=== CONT  TestAccAWSVpnGateway_basic
--- PASS: TestAccAWSVpnGateway_reattach (435.76s)
--- PASS: TestAccAWSVpnGateway_basic (247.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	876.805s
```

ewbankkit commented:
More success with acceptance tests having cherry picked the commit from https://github.com/terraform-providers/terraform-provider-aws/pull/8576:
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_deprecated'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_deprecated -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_deprecated
=== PAUSE TestAccAwsDxGatewayAssociation_deprecated
=== CONT  TestAccAwsDxGatewayAssociation_deprecated
--- PASS: TestAccAwsDxGatewayAssociation_deprecated (1381.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1381.806s
```

ewbankkit commented:
This PR will now incorporate the cross-account Gateway Proposal acceptance work started in https://github.com/terraform-providers/terraform-provider-aws/pull/8455. These changes will be incorporated into the `aws_dx_gateway_association` resource rather than creating a new `aws_dx_cross_account_gateway_association` resource.

ewbankkit commented:
Cross-account Direct Connect gateway association changes from https://github.com/terraform-providers/terraform-provider-aws/pull/8455 have been incorporated and all acceptance tests are now passing. Note that I am including the commit from https://github.com/terraform-providers/terraform-provider-aws/pull/8576 to ensure that VGW associations are correctly cleaned up.

#### aws_dx_gateway_association_proposal

```console
$ AWS_ALTERNATE_ACCESS_KEY_ID=AAAAAAAAAAAAAAAAAAAA AWS_ALTERNATE_SECRET_ACCESS_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociationProposal_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 2 -run=TestAccAwsDxGatewayAssociationProposal_ -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociationProposal_deprecated
=== PAUSE TestAccAwsDxGatewayAssociationProposal_deprecated
=== RUN   TestAccAwsDxGatewayAssociationProposal_basicVpnGateway
=== PAUSE TestAccAwsDxGatewayAssociationProposal_basicVpnGateway
=== RUN   TestAccAwsDxGatewayAssociationProposal_basicTransitGateway
=== PAUSE TestAccAwsDxGatewayAssociationProposal_basicTransitGateway
=== RUN   TestAccAwsDxGatewayAssociationProposal_disappears
=== PAUSE TestAccAwsDxGatewayAssociationProposal_disappears
=== RUN   TestAccAwsDxGatewayAssociationProposal_allowedPrefixes
=== PAUSE TestAccAwsDxGatewayAssociationProposal_allowedPrefixes
=== CONT  TestAccAwsDxGatewayAssociationProposal_deprecated
=== CONT  TestAccAwsDxGatewayAssociationProposal_allowedPrefixes
--- PASS: TestAccAwsDxGatewayAssociationProposal_deprecated (71.43s)
=== CONT  TestAccAwsDxGatewayAssociationProposal_disappears
--- PASS: TestAccAwsDxGatewayAssociationProposal_allowedPrefixes (94.61s)
=== CONT  TestAccAwsDxGatewayAssociationProposal_basicTransitGateway
--- PASS: TestAccAwsDxGatewayAssociationProposal_disappears (58.27s)
=== CONT  TestAccAwsDxGatewayAssociationProposal_basicVpnGateway
--- PASS: TestAccAwsDxGatewayAssociationProposal_basicVpnGateway (60.52s)
--- PASS: TestAccAwsDxGatewayAssociationProposal_basicTransitGateway (146.13s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	240.815s
```

#### aws_dx_gateway_association

```console
$ AWS_ALTERNATE_ACCESS_KEY_ID=AAAAAAAAAAAAAAAAAAAA AWS_ALTERNATE_SECRET_ACCESS_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 2 -run=TestAccAwsDxGatewayAssociation_ -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== RUN   TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
=== CONT  TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== CONT  TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_deprecatedSingleAccount (1348.63s)
=== CONT  TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount (1885.66s)
=== CONT  TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount (2193.34s)
=== CONT  TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount (1657.42s)
=== CONT  TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount (1232.65s)
=== CONT  TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount (1235.44s)
=== CONT  TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount (1844.14s)
--- PASS: TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount (1847.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	6623.071s
```

ewbankkit commented:
Still TODO:
* Add CHANGELOG notes

ewbankkit commented on aws/resource_aws_dx_gateway_association.go:
```diff
@@ -111,24 +189,66 @@ func resourceAwsDxGatewayAssociationRead(d *schema.ResourceData, meta interface{
+	d.Set("dx_gateway_owner_account_id", assoc.DirectConnectGatewayOwnerAccount)
+
+	tgwAttachmentId := ""
+	if aws.StringValue(assoc.AssociatedGateway.Type) == directconnect.GatewayTypeTransitGateway {
```
We need to get the transit gateway attachment here to ensure that we wait for it's deletion when this resource is deleted.
See discussion: https://github.com/terraform-providers/terraform-provider-aws/issues/8490#issuecomment-490980638.
Maybe move this code plus `ec2TransitGatewayAttachmentStateRefresh()` and `waitForEc2TransitGatewayAttachmentDeletion()` below to `ec2_transit_gateway.go`?

ewbankkit commented:
Removing _WIP_ as now ready for review,

ewbankkit commented:
Rebased to remove the fixes incorporated in https://github.com/terraform-providers/terraform-provider-aws/pull/8576.

aleksap commented:
Hi @ewbankkit ! So looks like you have code ready, it just needs to be merged?

ewbankkit commented:
@aleksap Functionality is complete, a review is required.

nywilken commented on aws/resource_aws_dx_gateway_association_proposal.go:
```diff
@@ -18,14 +20,47 @@ func resourceAwsDxGatewayAssociationProposal() *schema.Resource {
+					return false
+				}
+
+				if proposal != nil && aws.StringValue(proposal.ProposalState) == directconnect.GatewayAssociationProposalStateRequested {
```
looks like we can just return the results of the final condition here; dropping lines 35-39 for the following: 

```go
return proposal != nil && aws.StringValue(proposal.ProposalState) == directconnect.GatewayAssociationProposalStateRequested
```

ewbankkit commented:
@nywilken I have addressed all the review comments so far; Thanks.
RE https://github.com/terraform-providers/terraform-provider-aws/pull/8528#discussion_r284872722, we are only at step 1 of the `vpn_gateway_id` [attribute removal process](https://www.terraform.io/docs/extend/best-practices/deprecations.html#provider-attribute-removal) so there is not yet a migration function. However I renamed the test to `TestAccAwsDxGatewayAssociationProposal_VpnGatewayId` and this can be the basis for the migration test when that time comes.

ewbankkit commented on aws/resource_aws_dx_gateway_association.go:
```diff
@@ -1,19 +1,23 @@
 
 const (
 	gatewayAssociationStateDeleted = "deleted"
+
```
This isn't yet defined in `ec2/api.go`.

nywilken commented:
@ewbankkit I created a PR #8752 to handle the Transit Gateway error this resource is encountering during testing. Please try cherry-picking that change or rebasing as it has the latest master to verify things are working as expected. I tested things locally and all appears to be working as expected.

ewbankkit commented:
@nywilken That seems like a much better solution. I'll remove the transit gateway related stuff and test with your commit.

ewbankkit commented on website/docs/r/dx_gateway_association_proposal.html.markdown:
```diff
@@ -29,24 +35,37 @@ resource "aws_vpn_gateway" "example" {
-* `id` - Direct Connect Gateway Association Proposal identifier
+* `id` - Direct Connect Gateway Association Proposal identifier.
+* `associated_gateway_owner_account_id` - The ID of the AWS account that owns the VGW or transit gateway with which to associate the Direct Connect gateway.
+* `associated_gateway_type` - The type of the associated gateway, `transitGateway` or `virtualPrivateGateway`.
```
@nywilken You mean the `associated_gateway_type` attribute? I can see a possible use in a module scenario where `associated_gateway_id` is passed in and we don't know whether that refers to a VGW or DXGW (or some possible other future GW type) and dependent resource are created conditionally on the GW type.

ewbankkit commented on website/docs/r/dx_gateway_association.html.markdown:
```diff
@@ -61,20 +65,100 @@ resource "aws_dx_gateway_association" "example" {
 In addition to all arguments above, the following attributes are exported:
 
 * `id` - The ID of the Direct Connect gateway association resource.
+* `associated_gateway_type` - The type of the associated gateway, `transitGateway` or `virtualPrivateGateway`.
```
See my comments below - https://github.com/terraform-providers/terraform-provider-aws/pull/8528/files#r287051159.

ewbankkit commented on website/docs/r/dx_gateway_association.html.markdown:
````diff
@@ -61,20 +65,100 @@ resource "aws_dx_gateway_association" "example" {
+}
+```
+
+### VGW Cross-Account
````
I think this warrants a new example in the `examples` directory with a link from here and the `aws_dx_gateway_association_proposal` documentation. What do you think?

nywilken commented on website/docs/r/dx_gateway_association.html.markdown:
````diff
@@ -86,52 +86,7 @@ resource "aws_dx_gateway_association" "example" {
-  associated_gateway_owner_account_id = "${data.aws_caller_identity.creator.account_id}"
-}
-```
+A full example of how to to create a VPN Gateway in one AWS account, create a Direct Connect Gateway in a second AWS account, and associate the VPN Gateway with the Direct Connect Gateway via the `aws_dx_gateway_association_proposal` and `aws_dx_gateway_association` resources can be found in [the `./examples/dx-gateway-cross-account-vgw-association` directory within the Github Repository](https:/ ...
````
~to~

nywilken commented on website/docs/r/dx_gateway_association_proposal.html.markdown:
````diff
@@ -13,40 +13,15 @@ Manages a Direct Connect Gateway Association Proposal, typically for enabling cr
-}
 ```
 
+A full example of how to to create a VPN Gateway in one AWS account, create a Direct Connect Gateway in a second AWS account, and associate the VPN Gateway with the Direct Connect Gateway via the `aws_dx_gateway_association_proposal` and `aws_dx_gateway_association` resources can be found in [the `./examples/dx-gateway-cross-account-vgw-association` directory within the Github Repository](https:/ ...
````
ditto

nywilken commented on examples/dx-gateway-cross-account-vgw-association/README.md:
```diff
@@ -0,0 +1,18 @@
+# Direct Connect Gateway Cross-Account VGW Association
```
:100: Good call on moving this out to the examples directory.

ewbankkit commented:
Rebased from **master** with fixes from https://github.com/terraform-providers/terraform-provider-aws/pull/8752.

ewbankkit commented:
Acceptance tests:

```console
$ AWS_ALTERNATE_ACCESS_KEY_ID=AAAAAAAAAAAAAAAAAAAA AWS_ALTERNATE_SECRET_ACCESS_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 2 -run=TestAccAwsDxGatewayAssociation_ -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== RUN   TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
=== PAUSE TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
=== RUN   TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
=== CONT  TestAccAwsDxGatewayAssociation_deprecatedSingleAccount
=== CONT  TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_deprecatedSingleAccount (1356.21s)
=== CONT  TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_multiVpnGatewaysSingleAccount (1356.50s)
=== CONT  TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicTransitGatewaySingleAccount (1760.16s)
=== CONT  TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount
--- FAIL: TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount (1762.87s)
    testing.go:568: Step 0 error: errors during apply:
        
        Error: error waiting for Direct Connect gateway association (ga-xxxxxxxx) to become available: timeout while waiting for state to become 'associated' (last state: 'associating', timeout: 15m0s)
        
          on /var/folders/4j/lf5jdq5j4y91g80ytf6qzjwmk6drp5/T/tf-test232917661/main.tf line 39:
          (source code not available)
        
        
=== CONT  TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicVpnGatewaySingleAccount (1966.67s)
=== CONT  TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicVpnGatewayCrossAccount (1971.17s)
=== CONT  TestAccAwsDxGatewayAssociation_allowedPrefixesVpnGatewaySingleAccount
panic: test timed out after 2h0m0s
```

nywilken commented:
Running the test by itself appears to work properly. Have you tried updating the timeout interval to see if all tests pass without timeout? I'm currently running that now with a 30 minute timeout. 


```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== PAUSE TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
=== CONT  TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount
--- PASS: TestAccAwsDxGatewayAssociation_basicTransitGatewayCrossAccount (1480.06s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1480.093s
```

In the meantime, I am looking to see if there are any missing pieces as @bflad pointed me to this note `The transit gateway must be attached to a VPC or VPN in your AWS account.` [here](https://docs.aws.amazon.com/directconnect/latest/UserGuide/multi-account-associate-tgw.html#multi-account-tgw-create-proposal)

bflad commented:
This has been released in [version 2.12.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2120-may-24-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

aleksap commented:
@ewbankkit @nywilken 
Hello, 

Thank you for your work on this feature. 
I have noticed that on aws_dx_gateway_association resource, argument vpn_gateway_id  was deprecated in favor of associated_gateway_id. 
We have tested that and it works fine. 
However, when it comes to destroying those resources we get following messages:

Error: module.dx_vgw_assoc.aws_dx_gateway_association.dx_gw_assoc: **"vpn_gateway_id": required field is not set**
Error: module.dx_vgw_assoc.aws_dx_gateway_association.dx_gw_assoc: : **invalid or unknown key: associated_gateway_id** 

It almost looks like that "destroy" part is not aware that vpn_gateway_id is deprecated and is not recognizing associated_gateway_id argument. 

Please let me know if this makes sense and if it does should I open bug ticket. 

Thank you, 
Alex

ewbankkit commented:
@aleksap Strange. It looks like somehow an earlier version of the provider (pre-v2.12.0) is being run here, hence the errors about missing `vpn_gateway_id` and extraneous `associated_gateway_id`.
Could you add a sample of Terraform code that is exhibiting this behavior?
Thanks.

aleksap commented:
@ewbankkit I am not sure what happened but after deleting my .terraform folder and re-downloading provider it just started working!
Now, I'm pretty confident I didn't have .terraform before in that directory and can't really explain why it was doing this. 
I will keep an eye for it after next build/destroy cycle and let you know if we see the same. 

Thanks for the quick reply!
